crates/api/src/events.rs

```diff
@@ -0,0 +1,99 @@
+/// We are using tokio::sync::broadcast to support multiple connections via WebSocket.
+/// The idea is, that if two clients ask for the same stream of data, you don't wanna query it twice.
+/// Instead you listen on different task (See: `tokio::spawn` in `EventSubscriber::new`) and then send message to broadcast channel.
+/// Each websocket client has its own Receiver.
+/// Thanks to that we are not only reusing connection, but also limit dangerous `consumer.leak()` usage.
+use crate::config::KafkaConfig;
+use futures::task::{Context as FutCtx, Poll};
+use futures::{Future, Stream, StreamExt, TryStreamExt};
+use juniper::FieldResult;
+use std::pin::Pin;
+use tokio::sync::broadcast;
+use utils::messaging_system::consumer::CommonConsumer;
+use utils::messaging_system::Error;
+
+// TODO: Rename it to generic `Event` after adding support to RabbitMQ
+/// Owned generic message received from kafka.
+#[derive(Clone, Debug)]
+pub struct KafkaEvent {
+    pub key: Option<String>,
+    pub payload: Option<String>,
+}
+
+/// Wrapper to prevent accidental sending data to channel. `Sender` is used only for subscription mechanism
+pub struct EventSubscriber(broadcast::Sender<Result<KafkaEvent, Error>>);
+
+// We are using Box<dyn> approach (recommended) by Tokio maintainers,
+// as unfortunately `broadcast::Receiver` doesn't implement `Stream` trait,
+// and it is hard to achieve it without major refactor. Therefore we are using `async_stream` as a loophole.
+pub struct EventStream {
+    inner: Pin<Box<dyn Stream<Item = FieldResult<KafkaEvent>> + Send + Sync>>,
+}
+
+impl EventSubscriber {
+    /// Connects to kafka and sends all messages to broadcast channel.
+    pub async fn new<F, Fut>(
+        config: &KafkaConfig,
+        topic: &str,
+        on_close: F,
+    ) -> Result<(Self, EventStream), anyhow::Error>
+    where
+        F: FnOnce(String) -> Fut + Send + 'static,
+        Fut: Future<Output = ()>,
+    {
+        let (tx, rx) = broadcast::channel(32);
+
+        log::debug!("Create new consumer for topic: {}", topic);
+
+        let consumer =
+            CommonConsumer::new_kafka(&config.group_id, &config.brokers, &[topic]).await?;
+        let consumer = Box::leak(Box::new(consumer));
```

Comment:
```suggestion
        let consumer = consumer.leak();
```
Easier to trace(remove leaks once it's possible).

Reply:
I say more, we can't leak here, every new subscription will allocate new Consumer that will live forever

Reply:
![image](https://user-images.githubusercontent.com/1165825/105023199-f2a65980-5a4a-11eb-9616-875190f61502.png)
You dont have to worry about `leak` if there is no `leak` ;-)
`Subscriber` architecture uses `broadcast` which allowed me to remove the leak and have a normal local variable.